### PR TITLE
[AutoTuner] Add auto tuner to obtain optima configuration

### PR DESCRIPTION
**Pull request by Caozhou1995**
### PR types

New features
### PR changes

Others
### Description

Pcard-72023

The optimal configuration for large model with distributed training/inference often requires designing multiple sets of experiments based on experiences (network, parameter size, gpu memory or flops, etc.), and comparing the results to determine the optimal configuration. This process heavily relies on human experience, and the determined optimal configuration may not be the global optimal configuration. When any condition changes, the above process needs to be repeated repeatedly, resulting in poor usability of large models. 

To address the above issues, we have implemented AutoTuner based on Profiling, with the main modules as follows:
1. Provide clear json configuration for users to directly use AutoTuner, avoiding additional coding work for users
2. launch multi tasks one by one and automatically schedule and monitor.
3. Implement search module and pruning module,  support multiple search algorithms and pruning strategies.

At present, we have built-in grid search support for 8 dimensions, including dp degree, mp degree, pp degree, mbs, sharding degree, sharding stage, recompute, and recompute granularity. The example JSON is as follows:
image

The usage is as follows:
python -m paddle.distributed.launch  --devices "0,1,2,3,4,5,6,7" --auto_tuner_json=test.json  your_train.py   your_args

NOTE: Since the auto_tuner is non-invasive, users need to expose args in their script to enable the configuration generated by auto_tuner be executed.

### CODE REVIEW

**Review by XieYunshen — APPROVED** [review]
LGTM for `set_tests_properties(test_auto_tuner PROPERTIES LABELS "RUN_TYPE=EXCLUSIVE"
                                                  TIMEOUT 100)`

File: test/auto_parallel/test_auto_tuner.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+                tmp_dir.name,
+                "--auto_tuner_json",
+                test_json_path,
+                launch_model_path,
+            ]
+        )
+
+        process = subprocess.Popen(cmd)
+        process.wait()
+        self.assertEqual(process.returncode, 0)
```
Comment (zhiqiu) [inline test/auto_parallel/test_auto_tuner.py:90]: Check the config searched?

File: python/paddle/distributed/launch/main.py [hunk 2]
```diff
@@ -293,6 +293,98 @@ def launch():
...
+    elif ctx.is_auto_tuner_mode():
+        import copy
+        import json
+        import signal
+        import sys
+        import time
+
+        from ..auto_tuner.tuner import AutoTuner
+        from ..auto_tuner.utils import gen_new_args
+        from . import controllers
```
Comment (zhiqiu) [inline python/paddle/distributed/launch/main.py:305]: Better import at the top of the file

File: python/paddle/distributed/launch/main.py [hunk 3]
```diff
@@ -293,6 +293,98 @@ def launch():
...
+            gpus_per_node = len(ctx.args.devices.split(","))
+        tuner_cfg["nodes"] = int(ctx.args.nnodes)
+        tuner_cfg["num_gpus"] = gpus_per_node * tuner_cfg["nodes"]
+
+        # build AutoTuner to get new config
+        auto_tuner = AutoTuner(tuner_cfg)
+        cur_cfg = auto_tuner.search_once()
+
+        # get max time per task run
+        max_time_per_task = tuner_cfg.get("max_time_per_task", 1800)
```
Comment (zhiqiu) [inline python/paddle/distributed/launch/main.py:335]: max_time_per_task -> max_time_in_seconds_per_task?

File: python/paddle/distributed/auto_tuner/tuner.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    The AutoTuner can automatically provide running task based on user-defined settings
+    and the task will be launched for execution.
+
+    Args:
+        tuner_cfg (dict): The configuration of auto tuner user defined.
+    """
+
+    def __init__(self, tuner_cfg):
+        self.cur_task_id = 1
+        self.task_limit = tuner_cfg.get("task_limit", 100)
```
Comment (zhiqiu) [inline python/paddle/distributed/auto_tuner/tuner.py:30]: DEFAULT_MAX_TASK_LIMIT = 100 ?

**Review by zhiqiu — APPROVED** [review]
LGTM, u can refine code with the comments in the next pr.

**Pull request merged**